esp-hal/src/ledc/timer.rs

```diff
@@ -97,150 +159,474 @@ pub mod config {
+            .conf()
+            .modify(|_, w| unsafe { w.div_num().bits(divisor) });
+    }
 
-/// Interface for HW configuration of timer
-pub trait TimerHW<S: TimerSpeed> {
-    /// Get the current source timer frequency from the HW
-    fn get_freq_hw(&self) -> Option<HertzU32>;
+    fn set_pause_bit(number: Number, pause: bool) {
+        Ledc::register_block()
+            .hstimer(number as usize)
+            .conf()
+            .modify(|_, w| w.pause().bit(pause));
+    }
 
-    /// Configure the HW for the timer
-    fn configure_hw(&self, divisor: u32);
+    fn set_reset_bit(number: Number, reset: bool) {
+        Ledc::register_block()
+            .hstimer(number as usize)
+            .conf()
+            .modify(|_, w| w.rst().bit(reset));
+    }
+
+    fn param_update(_: Number) {
+        // The parameters will be updated in the next cycle
+    }
 
-    /// Update the timer in HW
-    fn update_hw(&self);
+    fn get_counter(number: Number) -> u32 {
+        Ledc::register_block()
+            .hstimer(number as usize)
+            .value()
+            .read()
+            .cnt()
+            .bits() as _
+    }
 }
 
-/// Timer struct
-pub struct Timer<'a, S: TimerSpeed> {
-    ledc: &'a crate::peripherals::ledc::RegisterBlock,
+/// Builder for Timer
+pub struct TimerBuilder<S: TimerSpeed> {
+    phantom: PhantomData<S>,
     number: Number,
-    duty: Option<config::Duty>,
-    frequency: u32,
-    configured: bool,
-    use_ref_tick: bool,
-    clock_source: Option<S::ClockSourceType>,
 }
 
-impl<'a, S: TimerSpeed> TimerIFace<S> for Timer<'a, S>
-where
-    Timer<'a, S>: TimerHW<S>,
-{
-    /// Return the frequency of the timer
-    fn get_freq(&self) -> Option<HertzU32> {
-        self.get_freq_hw()
+impl<S: TimerSpeed> TimerBuilder<S> {
+    /// Create a new instance of a timer builder
+    /// `number`, `duty`, `clock_source` and `frequency` are the required
+    /// parameters. other parameters can be set by calling the corresponding
+    /// methods.
+    pub fn new(
+        _ledc: &Ledc<'_>,
+        number: Number,
+        duty: Duty,
+        clock_source: S::ClockSourceType,
+        frequency: HertzU32,
+    ) -> Result<TimerBuilder<S>, Error> {
+        S::set_duty(number, duty);
+        let builder = TimerBuilder {
+            phantom: PhantomData,
+            number,
+        };
+        builder.set_frequency(duty, clock_source, frequency)
+    }
+
+    /// Create a new instance of a timer builder from a configured timer
+    pub fn from_configured_timer(timer: Timer<S>) -> TimerBuilder<S> {
+        TimerBuilder {
+            phantom: PhantomData,
+            number: timer.number,
+        }
     }
 
-    /// Configure the timer
-    fn configure(&mut self, config: config::Config<S::ClockSourceType>) -> Result<(), Error> {
-        self.duty = Some(config.duty);
-        self.clock_source = Some(config.clock_source);
+    /// Set the frequency of the timer
+    pub fn set_frequency(
```

Comment:
Since this takes `duty` and `clock_source`, this function seems to do a lot more than just adjusting the frequency.